### PR TITLE
Replicate stack gan

### PR DESCRIPTION
**Pull request by Devin-Taylor**
- bugfix: loss function sigmoid
- padding for convs
- custom weight initialisers
- create embedding which is mean of random samples drawn from 10 options
- converted dataset downloading to use a local copy of raw data if available, otherwise create it
- cropping preprocessing functionality for birds, cropping based on bounding boxes provided

### CODE REVIEW

File: models/conditional_gans.py [hunk 1]
```diff
@@ -116,9 +116,8 @@ def generate_conditionals(self, embedding):
         """
         # NOTE embedding is dim (batch, 10, 1024) where 10 is different samples for same image. Options are to either
         # flatten all features or average across embeddings
-        x = embedding[:, 0, :] # rather just select one of the embeddings
-        mean = self.lrelu1(self.dense_mean(x))
-        sigma = self.lrelu2(self.dense_sigma(x))
+        mean = self.lrelu1(self.dense_mean(embedding))
+        sigma = self.lrelu2(self.dense_sigma(embedding))
```
Comment (alecokas) [inline models/conditional_gans.py:120]: Which dimension do `dense_mean` and `dense_sigma` operate over?

File: models/discriminators.py [hunk 2]
```diff
@@ -67,36 +67,40 @@ def __init__(self, img_size, kernel_size, num_filters, lr):
...
         super().__init__()
+
+        self.w_init = tf.random_normal_initializer(stddev=0.02)
+        self.bn_init = tf.random_normal_initializer(1., 0.02)
+
         self.img_size = img_size
         num_channels = self.img_size[0]
         # self.xent_loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=True)
-        self.optimiser = tf.keras.optimizers.Adam(lr)
+        self.optimiser = tf.keras.optimizers.Adam(lr, beta_1=0.5)
```
Comment (alecokas) [inline models/discriminators.py:77]: Is `0.5` the number from the paper? The standard Adam `beta_1` is 0.9

File: models/generators.py [hunk 3]
```diff
@@ -76,57 +76,67 @@ def __init__(self, img_size, num_latent_dims, kernel_size, num_filters,
                 lr : float
         """
         super().__init__()
+
+        self.w_init = tf.random_normal_initializer(stddev=0.02)
+        self.bn_init = tf.random_normal_initializer(1., 0.02)
```
Comment (alecokas) [inline models/generators.py:81]: We probably need to find out what the best initialization for GANs are if this isn't from the paper

File: models/generators.py [hunk 4]
```diff
@@ -76,57 +76,67 @@ def __init__(self, img_size, num_latent_dims, kernel_size, num_filters,
...
+
         self.img_size = img_size
 
         num_output_channels = self.img_size[0]
         self.num_latent_dims = num_latent_dims
         # self.xent_loss_fn = tf.keras.losses.BinaryCrossentropy(from_logits=True)
-        self.optimiser = tf.keras.optimizers.Adam(lr)
+        self.optimiser = tf.keras.optimizers.Adam(lr, beta_1=0.5)
```
Comment (alecokas) [inline models/generators.py:88]: Same as above

File: utils/data_helpers.py [hunk 5]
```diff
@@ -56,6 +57,18 @@ def create_tfrecords(dataset_type, tfrecords_dir, image_source_dir, text_source_
     else:
         raise Exception(f'{dataset_type} is not a recognised dataset type')
 
+def extract_image_bounding_boxes(image_filenames, base_path='data/CUB_200_2011_with_text/images/CUB_200_2011'):
+    '''
+    Returns a map of filename to bounding box in format [x-top, y-top, w, h]
+    '''
```
Comment (alecokas) [inline utils/data_helpers.py:63]: Use double quotes on docstrings to be consistent

File: models/conditional_gans.py (hunk 1 above)
Comment (Devin-Taylor) [reply models/conditional_gans.py:120]: The embedding dimension (1024)

File: models/discriminators.py (hunk 2 above)
Comment (Devin-Taylor) [reply models/discriminators.py:77]: Yup, taken from the paper. Just wanted to get it training properly so replicated as much as possible.

File: models/generators.py (hunk 3 above)
Comment (Devin-Taylor) [reply models/generators.py:81]: Taken from the paper, too

File: models/generators.py (hunk 4 above)
Comment (Devin-Taylor) [reply models/generators.py:88]: Again, from paper

File: utils/data_helpers.py (hunk 5 above)
Comment (Devin-Taylor) [reply utils/data_helpers.py:63]: Haha my bad, I'll change it in next PR
